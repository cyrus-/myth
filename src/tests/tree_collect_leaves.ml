let s = "
type bool =
  | True
  | False

type tree =
  | Leaf
  | Node of tree * bool * tree

type list =
  | Nil
  | Cons of bool * list

let rec append (l1:list) (l2:list) : list =
  match l1 with
  | Nil -> l2
  | Cons (x, l1) -> Cons (x, append l1 l2)
;;

let tree_collect_leaves : tree -> list |>
  { Leaf => []
  | Node (Leaf, True, Leaf) => [True]
  | Node (Leaf, False, Leaf) => [False]
  | Node (Node (Leaf, True, Leaf), False, Leaf) => [True; False]
  | Node (Node (Leaf, False, Leaf), True, Leaf) => [False; True]
  | Node (Leaf, False, Node (Leaf, True, Leaf)) => [False; True]
  } = ?"
