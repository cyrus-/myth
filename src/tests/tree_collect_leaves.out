(* Generated Thu Nov 13 00:35:54 2014 *)
let tree_collect_leaves : tree -> list =
  let rec f1 (t1:tree) : list =
    match t1 with
      | Leaf -> Nil
      | Node (t2, b1, t3) -> append (f1 t2) (Cons (b1, f1 t3))
  in
    f1
;;
